pkg/lb/servicelb/manager.go

```diff
@@ -76,57 +90,220 @@ func (m *Manager) updateHealthCondition(uid string, isHealthy bool) error {
-		return err
+		return fmt.Errorf("ensure service failed, error: %w", err)
 	}
-	if err := m.ensureEndpointSlice(lb); err != nil {
+
+	eps, err := m.ensureEndpointSlice(lb)
+	if err != nil {
+		return fmt.Errorf("ensure endpointslice failed, error: %w", err)
+	}
+
+	m.ensureProbes(lb, eps)
+
+	return nil
+}
+
+func (m *Manager) DeleteLoadBalancer(lb *lbv1.LoadBalancer) error {
+	eps, err := m.endpointSliceCache.Get(lb.Namespace, lb.Name)
+	if err != nil {
 		return err
 	}
+	m.removeProbes(lb, eps)
 
-	return m.ensureProber(lb)
+	return nil
 }
 
-func (m *Manager) DeleteLoadBalancer(lb *lbv1.LoadBalancer) {
-	m.removeProber(lb)
+func (m *Manager) GetBackendServers(lb *lbv1.LoadBalancer) ([]pkglb.BackendServer, error) {
+	eps, err := m.endpointSliceCache.Get(lb.Namespace, lb.Name)
+	if err != nil {
+		return nil, err
+	}
+
+	servers := make([]pkglb.BackendServer, 0, len(eps.Endpoints))
+	for _, ep := range eps.Endpoints {
+		vmi, err := m.vmiCache.Get(ep.TargetRef.Namespace, ep.TargetRef.Name)
+		if err != nil {
+			return nil, err
+		}
+
+		servers = append(servers, &Server{vmi})
+	}
+
+	return servers, nil
 }
 
-func (m *Manager) ensureProber(lb *lbv1.LoadBalancer) error {
-	if lb.Spec.HeathCheck == nil {
-		m.removeProber(lb)
-		return nil
+func (m *Manager) AddBackendServers(lb *lbv1.LoadBalancer, servers []pkglb.BackendServer) error {
+	eps, err := m.endpointSliceCache.Get(lb.Namespace, lb.Name)
+	if err != nil {
+		return err
 	}
-	option := prober.HealthOption{
-		SuccessThreshold: lb.Spec.HeathCheck.SuccessThreshold,
-		FailureThreshold: lb.Spec.HeathCheck.FailureThreshold,
-		Timeout:          time.Duration(lb.Spec.HeathCheck.TimeoutSeconds) * time.Second,
-		Period:           time.Duration(lb.Spec.HeathCheck.PeriodSeconds) * time.Second,
-		InitialCondition: true,
+	var epsCopy *discoveryv1.EndpointSlice
+	endpoints := make([]*discoveryv1.Endpoint, 0, len(servers))
+	for _, server := range servers {
+		if flag, _, err := isExisting(eps, server); err != nil || flag {
+			continue
+		}
+
+		address, ok := server.GetAddress()
+		if !ok {
+			continue
+		}
+
+		endpoint := discoveryv1.Endpoint{
+			Addresses: []string{address},
+			TargetRef: &corev1.ObjectReference{
+				Namespace: server.GetNamespace(),
+				Name:      server.GetName(),
+				UID:       server.GetUID(),
+			},
+			Conditions: discoveryv1.EndpointConditions{
+				Ready: newTrue(),
+			},
+		}
+		if epsCopy == nil {
+			epsCopy = eps.DeepCopy()
+		}
+		epsCopy.Endpoints = append(epsCopy.Endpoints, endpoint)
+		endpoints = append(endpoints, &endpoint)
+	}
+
+	if epsCopy != nil {
+		if _, err := m.endpointSliceClient.Update(epsCopy); err != nil {
+			return err
+		}
+	}
+
+	if lb.Spec.HealthCheck != nil && lb.Spec.HealthCheck.Port != 0 {
+		for _, endpoint := range endpoints {
+			m.addOneProbe(lb, endpoint)
+		}
 	}
 
+	return nil
+}
+
+func (m *Manager) RemoveBackendServers(lb *lbv1.LoadBalancer, servers []pkglb.BackendServer) error {
```

Comment:
better to return true/false,  to know if some backend servers are really removed from this LB